File: Sorting and Searching/Towers/towers.py
# Gives TLE. TODO: Insertion and deletion in list is O(n). Use a data structure with O(logn) insertion and deletion.

from sys import stdin
input = stdin.readline

import time

class MultiSet:
    def __init__(self):
        self.data = []
        self.size = 0
        
    def upper_bound(self, arr, k):
        l, r = 0, len(arr)
        while l < r:
            m = (l + r) // 2
            if arr[m] <= k:
                l = m + 1
            else:
                r = m
        return l
    
    def insert(self, x):
        idx = self.upper_bound(self.data, x)
        self.data.insert(idx, x)
        self.size += 1
    
    def erase(self, x):
        if x in self.data:
            self.data.remove(x)
            self.size -= 1
    

n = int(input())
towers = list(map(int, input().split()))
start = time.time()
temp = MultiSet()
for i in range(n):
    it = temp.upper_bound(temp.data, towers[i])
    if it == temp.size:
        temp.insert(towers[i])
    else:
        temp.erase(temp.data[it])
        temp.insert(towers[i])
end = time.time()
print(temp.size)

print("Time taken: ", end-start, "s")
# RESULT: TLE

File: Sorting and Searching/Towers/towers.cpp
#include<bits/stdc++.h>
using namespace std;

/* TYPES  */
#define ll long long
#define pii pair<int, int>
#define pll pair<long long, long long>
#define vi vector<int>
#define vll vector<long long>
#define mii map<int, int>
#define si set<int>
#define sc set<char>

/* FUNCTIONS */
#define f(i,s,e) for(long long int i=s;i<e;i++)
#define cf(i,s,e) for(long long int i=s;i<=e;i++)
#define rf(i,e,s) for(long long int i=e-1;i>=s;i--)
#define pb push_back
#define eb emplace_back


/* PRINTS */
template <class T>
void print_v(vector<T> &v) { cout << "{"; for (auto x : v) cout << x << ","; cout << "\b}"; }

/* UTILS */
#define MAXN 100005
#define BIGINT 1000000009

bool cmp(pii a, pii b){
    return a.first < b.first;
}

int main(){
    ios_base::sync_with_stdio(0);
    cin.tie(0); cout.tie(0);

    int n;
    cin>>n;
    vi towers(n);
    f(i,0,n) cin>>towers[i];

    multiset<int> temp;
    ll curr = n;
    f(i,0,n){
        auto it = temp.upper_bound(towers[i]);
        if (it != temp.end()) {
            temp.erase(it);
            curr--;
        }
        temp.insert(towers[i]);
    }
    cout<<curr<<endl;

    cerr << "time taken : " << (float)clock() / CLOCKS_PER_SEC << " secs" << endl;
    return 0;
}
// Result: ACCEPTED